pkg/controller/seed-controller-manager/mla/rulegroup_controller.go

```diff
@@ -0,0 +1,353 @@
+				NamespacedName: types.NamespacedName{
+					Name:      ruleGroup.Name,
+					Namespace: ruleGroup.Namespace,
+				},
+			})
+		}
+		return requests
+	})
+
+	clusterPredicate := predicate.Funcs{
+		// For Update event, only trigger reconciliation when MonitoringEnabled changes.
+		UpdateFunc: func(event event.UpdateEvent) bool {
+			oldCluster := event.ObjectOld.(*kubermaticv1.Cluster)
+			newCluster := event.ObjectNew.(*kubermaticv1.Cluster)
+			oldMonitoringEnabled := oldCluster.Spec.MLA != nil && oldCluster.Spec.MLA.MonitoringEnabled
+			newMonitoringEnabled := newCluster.Spec.MLA != nil && newCluster.Spec.MLA.MonitoringEnabled
+			return oldMonitoringEnabled != newMonitoringEnabled
+		},
+	}
+	if err := c.Watch(&source.Kind{Type: &kubermaticv1.Cluster{}}, enqueueRuleGroupsForCluster, clusterPredicate); err != nil {
+		return fmt.Errorf("failed to watch Cluster: %w", err)
+	}
+
+	return nil
+}
+
+func (r *ruleGroupReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := r.log.With("request", request)
+	log.Debug("Processing")
+
+	ruleGroup := &kubermaticv1.RuleGroup{}
+	if err := r.Get(ctx, request.NamespacedName, ruleGroup); err != nil {
+		return reconcile.Result{}, ctrlruntimeclient.IgnoreNotFound(err)
+	}
+
+	cluster := &kubermaticv1.Cluster{}
+	if err := r.Get(ctx, types.NamespacedName{Name: ruleGroup.Spec.Cluster.Name}, cluster); err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to get cluster: %w", err)
+	}
+
+	// Add a wrapping here so we can emit an event on error
+	result, err := kubermaticv1helper.ClusterReconcileWrapper(
+		ctx,
+		r.Client,
+		r.workerName,
+		cluster,
+		r.versions,
+		kubermaticv1.ClusterConditionMLAControllerReconcilingSuccess,
+		func() (*reconcile.Result, error) {
+			return r.ruleGroupController.reconcile(ctx, cluster, ruleGroup)
+		},
+	)
+	if err != nil {
+		r.log.Errorw("Failed to reconcile cluster", "cluster", cluster.Name, zap.Error(err))
+		r.recorder.Event(cluster, corev1.EventTypeWarning, "ReconcilingError", err.Error())
+	}
+	if result == nil {
+		result = &reconcile.Result{}
+	}
+	return *result, err
+
+}
+
+type ruleGroupController struct {
+	ctrlruntimeclient.Client
+	httpClient *http.Client
+
+	log            *zap.SugaredLogger
+	cortexRulerURL string
+}
+
+func newRuleGroupController(
+	client ctrlruntimeclient.Client,
+	log *zap.SugaredLogger,
+	httpClient *http.Client,
+	cortexRulerURL string,
+) *ruleGroupController {
+	return &ruleGroupController{
+		Client:         client,
+		httpClient:     httpClient,
+		log:            log,
+		cortexRulerURL: cortexRulerURL,
+	}
+}
+
+func (r *ruleGroupController) reconcile(ctx context.Context, cluster *kubermaticv1.Cluster, ruleGroup *kubermaticv1.RuleGroup) (*reconcile.Result, error) {
+	requestURL, err := r.getRequestURL(ruleGroup)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get request URL: %w", err)
+	}
+
+	if !ruleGroup.DeletionTimestamp.IsZero() {
```

Comment:
why do we need this condition? 
you can call `handleDeletion` in the next one

Reply:
This is need for the case that user only wants to delete a `RuleGroup`, not only because the cluster is being deleted or mla is disabled

Reply:
oh, I see, it's `DeletionTimestamp` on `ruleGroup` and the next one on the `cluster`